### ISSUE TITLE
Error deleting categories between releases 3.0.2 and 3.0.4

### ISSUE THREAD

**Issue by giovannipds** [rinvex/laravel-categories#64]
Hi! I've upgraded the package at my app and it started getting errors when deleting categories. I'd like to understand why. **At 3.0.2 it was all working good**.

It's generating error when `Category::destroy($id);` is called or `Category::findOrFail($id)->delete()`.

The exception shows something similar to:

```
{
    "class": "Illuminate\\Database\\Eloquent\\ModelNotFoundException",
    "message": "No query results for model [App\\Category].",
    "code": 0,
    "file": "C:\\Users\\user\\Desktop\\admin\\vendor\\laravel\\framework\\src\\Illuminate\\Database\\Eloquent\\Builder.php:470",
    "trace": [
        "C:\\Users\\user\\Desktop\\admin\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\SerializesAndRestoresModelIdentifiers.php:102",
        "C:\\Users\\user\\Desktop\\admin\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\SerializesAndRestoresModelIdentifiers.php:57",
        "C:\\Users\\user\\Desktop\\admin\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\SerializesModels.php:45",
        "C:\\Users\\user\\Desktop\\admin\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\CallQueuedHandler.php:171",
        "C:\\Users\\user\\Desktop\\admin\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\Jobs\\Job.php:203",
        "C:\\Users\\user\\Desktop\\admin\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\Jobs\\Job.php:182",
        "C:\\Users\\user\\Desktop\\admin\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\CallQueuedHandler.php:157",
        "C:\\Users\\user\\Desktop\\admin\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\CallQueuedHandler.php:56",
        "C:\\Users\\user\\Desktop\\admin\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\Jobs\\Job.php:88",
        "C:\\Users\\user\\Desktop\\admin\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\SyncQueue.php:45",
        "C:\\Users\\user\\Desktop\\admin\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\Queue.php:44",
        "C:\\Users\\user\\Desktop\\admin\\vendor\\laravel\\framework\\src\\Illuminate\\Broadcasting\\BroadcastManager.php:127",
        "C:\\Users\\user\\Desktop\\admin\\vendor\\laravel\\framework\\src\\Illuminate\\Events\\Dispatcher.php:2
    ]
}
```

**App\Category.php**
```php
<?php

namespace App;

use App\Movie;
use App\Traits\Activable;
use Rinvex\Categories\Models\Category as RinvexCategory;
use Spatie\MediaLibrary\HasMedia\HasMedia;
use Spatie\MediaLibrary\HasMedia\HasMediaTrait;

class Category extends RinvexCategory implements HasMedia
{
    use Activable, HasMediaTrait;

    protected $appends = [
        'image_url',
    ];

    protected $cacheLifetime = 0;

    protected $fillable = [
        'active',
        'name',
        'parent_id',
        'private',
    ];

    public function getImageAttribute()
    {
        return $this->getFirstMedia('image');
    }

    public function getImageUrlAttribute()
    {
        return $this->image ? $this->image->getFullUrl() : null;
    }

    public function movies()
    {
        return $this->entries(Movie::class);
    }

    public function registerMediaCollections()
    {
        $this
            ->addMediaCollection('image')
            ->acceptsMimeTypes(['image/jpeg'])
            ->singleFile()
            ->useDisk('categories');
    }
}
```

**CategoryController.php**
```php
<?php

namespace App\Http\Controllers;

use App\Category;
use Illuminate\Http\Request;

class CategoryController extends Controller
{
    public function destroy(Request $request, int $id)
    {
        Category::destroy($id);
        return redirect('admin/categories')
            ->with('deleted', __("Category has been deleted"));
    }
}
```

Any tips on why is this happening?

**Comment by mbryne** [comment]
I'm also experiencing this issue, with a similar setup where I have a custom model that extends `Rinvex\Categories\Models\Category`.

```php
// this works, direct
Category::find(2)->delete();

// throws ModelNotFoundException
DummyCategory::find(2)->delete();

// dummy class
class DummyCategory extends \Rinvex\Categories\Models\Category
{
}
```

When debugging the issue, it appears this comes from dispatching the CategoryDeleted event? 
I am (wildly) assuming that the SerializesModels trait could be expecting the main Category model?

I was able to resolve the issue by disabling the two category events:

```php
class DummyCategory extends \Rinvex\Categories\Models\Category
{

    /**
     * The event map for the model.
     *
     * @var array
     */
    protected $dispatchesEvents = [
    ];
    
}

```

I'm not sure of the side effects of disabling those events but I hope that helps.

EDIT: Cascading deletion is effected by disabling those events, so child categories don't get deleted with the events deleted, not ideal or much a solution sorry

**Comment by giovannipds** [comment]
@mbryne

**Comment by giovannipds** [comment]
@mbryne just to be sure, do you have your custom model inserted at `config/rinvex.categories.php`?

```php
<?php

declare(strict_types=1);

return [

    // Categories Database Tables
    'tables' => [

        'categories' => 'categories',
        'categorizables' => 'categorizables',

    ],

    // Categories Models
    'models' => [
        // 'category' => \Rinvex\Categories\Models\Category::class,
        'category' => \App\DummyCategory::class,
    ],

];
```

**Comment by giovannipds** [comment]
I remember this change solved some issues to me.

**Comment by mbryne** [comment]
I use multiple extended classes that restrict the scope:

```php
class DummyCategory extends \Rinvex\Categories\Models\Category
{
 
    protected static function boot()
    {
            parent::boot();

            static::addGlobalScope('type', function (Builder $builder) {
                $builder->where('type', '=', 'dummy');
            });

            static::creating(function ($category) {
                $category->type = 'dummy';
            });
        }

    }
}
```

This wasn't the issue on this end though, even without the boot / code above issue occurs with the serialization / deserialization classes in the stack trace above because it's expecting a Category and it received a DummyCategory, i.e.:

```  
"C:\\Users\\user\\Desktop\\admin\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\SerializesModels.php:45",
```

The issue seems to be related to people extending the functionality using their own inherited models rather than setting the config entry you have outlined. I know that is how the package has been designed, overiding that single config variable if you use another model but I was keen to have multiple category models etc. Not sure if this is resolvable without something like this: https://github.com/rinvex/laravel-categories/issues/24

**Comment by mbryne** [comment]
Just an update on this, what I've found is that having that inherited model and global scope concept was causing too many other headaches so I have stuck with the single model approach recommended above.

What I am doing for my custom models above is extending from another BaseCategory class that applies the global scopes etc to that. I then use the main app('rinvex.categories.category') model for any nested tree actions (moving, deleting etc)

This means I can no longer produce the error mentioned by @giovannipds above so I won't be able to assist with the issue any further. 

Thanks for the suggestions and the efforts put into the plugin.

**Comment by Omranic** [comment]
The `CategoryDeleted` event issue have been solved, please check the recent releases v4.x and let me know if you still have any issues, feel free to reopen this issue again. Thanks

**Comment by sashsvamir** [comment]
on 4.1.0 i'm still have this issue:
```php
// action in controller:
public function destroy(Category $category) {
    $category->delete(); // this leads to `NotFoundHttpException`
    // ...
}
```
this happens whan called `$this->fireModelEvent('deleted', false);` in Eloquent `Model.php`

**Comment by dhi2** [comment]
I have the same issue too on 4.0.

**Comment by dhi2** [comment]
@Omranic I have the same issue too on 4.0.

**Comment by dhi2** [comment]
For anyone still having this issue, I found a work around.

In my case I am integrating this package with Nova, so I would define the resource as such:

```php
use Rinvex\Categories\Models\Category;

class Interest extends Resource
{
    public static $model = Category::class;
   ...

```

I can create categories normally, but when trying to delete I would get the error `No query results for model ...`. To solve the problem now my Nova delete action looks something like this:
```php
namespace App\Nova\Actions;

use Exception;
use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;
use Illuminate\Support\Collection;
use Laravel\Nova\Actions\DestructiveAction;
use Laravel\Nova\Fields\ActionFields;
use \Rinvex\Categories\Models\Category;

class DeleteInterest extends DestructiveAction
{
    use InteractsWithQueue, Queueable, SerializesModels;

    public $withoutActionEvents = true;

    public function handle(ActionFields $fields, Collection $models)
    {
        foreach ($models as $model) {
            $cat = Category::find($model->id);

            if($cat){
                try {
                    $cat->children()->delete();
                    $cat->delete();
                } catch (Exception $e) {
                }
            }

            try {
                $model->children()->delete();
                $model->delete();
            } catch (Exception $e) {
            }
        }
    }

    public function fields()
    {
        return [];
    }
}
```

The idea is first I try to delete the category by retrieving the model, then from `Collection`, one of which will work.

**Comment by userbig** [comment]
Why this issue is closed? I Have the same issue on 4.1.1v (PHP 7.4.12)

```
>>> $s = app('rinvex.categories.category')->find(1)
=> Rinvex\Categories\Models\Category {#3433
     id: 1,
      ...
   }
>>> $s->delete();
Illuminate/Database/Eloquent/ModelNotFoundException with message 'No query results for model [Rinvex/Categories/Models/Category].'
`
```

**Comment by Omranic** [comment]
This has been fixed in these two commits: 8cffab84d9cebb316bcc1d837704d47e180658c6 & 5e1f8d12bafbf78fd3bdbf27df47e7e5b9f34de4
